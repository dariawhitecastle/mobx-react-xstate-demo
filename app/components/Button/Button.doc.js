import { describe, PropTypes } from 'react-desc'

import { genericProps, colorPropType } from '../../utils/propTypes'

export const doc = Button => {
  const DocumentedButton = describe(Button)
    .description('A button.')
    .details(
      `You can provide a single function child that will be called with
      'hover' and 'focus' keys. This allows you to customize the rendering
      of the Button in those cases.`,
    )
    .usage(
      `import { Button } from 'components/Button';
<Button primary label='Label' />`,
    )
    .intrinsicElement('button')

  DocumentedButton.propTypes = {
    ...genericProps,
    active: PropTypes.bool.description('Whether the button is active.').defaultValue(false),
    color: colorPropType.description(
      'Fill color for primary, label color for plain, border color otherwise.',
    ),
    disabled: PropTypes.bool.description('Whether the button is disabled.').defaultValue(false),
    fill: PropTypes.oneOfType([PropTypes.oneOf(['horizontal', 'vertical']), PropTypes.bool])
      .description('Whether the button expands to fill all of the available width and/or height.')
      .defaultValue(false),
    focusIndicator: PropTypes.bool
      .description("Whether when 'plain' it should receive a focus outline.")
      .defaultValue(true),
    hoverIndicator: PropTypes.oneOfType([
      PropTypes.bool,
      PropTypes.string,
      PropTypes.oneOf(['background']),
      PropTypes.shape({
        background: PropTypes.oneOfType([PropTypes.bool, PropTypes.string]),
      }),
    ])
      .description(
        `The hover indicator to apply when the user is mousing over the
button. An object can be also be specified for color index support:
{background: 'neutral-2'}. This prop is meant to be used only
with plain Buttons.`,
      )
      .defaultValue(false),
    href: PropTypes.string.description('If specified, the button will behave like an anchor tag.'),
    target: PropTypes.oneOf(['_self', '_blank', '_parent', '_top']).description(
      `Specifies where to display the URL defined in the href property.`,
    ),
    icon: PropTypes.element.description('Icon element to place in the button.'),
    gap: PropTypes.oneOfType([
      PropTypes.oneOf(['xxsmall', 'xsmall', 'small', 'medium', 'large', 'xlarge']),
      PropTypes.string,
    ])
      .description(`The amount of spacing between icon and label in the button.`)
      .defaultValue('small'),
    label: PropTypes.node.description('Label text to place in the button.'),
    onClick: PropTypes.func.description(
      `Click handler. Not setting this property and not specifying a href
causes the Button to be disabled.`,
    ),
    plain: PropTypes.bool
      .description(
        `Whether this is a plain button with no border or pad.
Non plain button will show both pad and border.
The plain button has no border and unless the icon prop exist it has no pad as well.`,
      )
      .defaultValue(false),
    primary: PropTypes.bool
      .description(
        'Whether this is a primary button. There should be at most one per page or screen.',
      )
      .defaultValue(false),
    reverse: PropTypes.bool
      .description(
        `Whether an icon and label should be reversed so that the icon is at the
end of the anchor.`,
      )
      .defaultValue(false),
    type: PropTypes.oneOf(['button', 'reset', 'submit'])
      .description('The type of button. Set the type to submit for the default button on forms.')
      .defaultValue('button'),
    as: PropTypes.oneOfType([PropTypes.string, PropTypes.func]).description(
      `The DOM tag or react component to use for the element.`,
    ),
  }

  return DocumentedButton
}

export const themeDoc = {
  'global.hover.color': {
    description: 'The background color when hovering.',
    type: 'string',
    defaultValue: "{ dark: 'white', light: 'black' }",
  },
  'global.edgeSize.small': {
    description: 'The padding around an icon-only button.',
    type: 'string',
    defaultValue: '12px',
  },
  'global.colors.control': {
    description: 'The color of the border.',
    type: 'string | { dark: string, light: string }',
    defaultValue: "{ dark: 'accent-1', light: 'brand', }",
  },
  'global.colors.brand': {
    description: 'The light version of the border.',
    type: 'string',
    defaultValue: '#7D4CDB',
  },
  'global.colors.text': {
    description: 'The color of the text label.',
    type: 'string | { dark: string, light: string }',
    defaultValue: "{ dark: '#f8f8f8', light: '#444444' }",
  },
  'text.medium.size': {
    description: 'The font size of the text label.',
    type: 'string',
    defaultValue: '18px',
  },
  'text.medium.height': {
    description: 'The line height of the text label.',
    type: 'string',
    defaultValue: '24px',
  },
  'button.border.color': {
    description: `The color of the border.`,
    type: 'string | { dark: string, light: string }',
  },
  'button.border.radius': {
    description: 'The corner radius.',
    type: 'string',
    defaultValue: '18px',
  },
  'button.border.width': {
    description: 'The border width.',
    type: 'string',
    defaultValue: '2px',
  },
  'button.color': {
    description: `The color of the text label.`,
    type: 'string | { dark: string, light: string }',
  },
  'button.primary.color': {
    description: `The color of the background for primary buttons.`,
    type: 'string | { dark: string, light: string }',
  },
  'button.disabled.opacity': {
    description: 'The opacity when the button is disabled.',
    type: 'number',
    defaultValue: 0.3,
  },
  'button.padding.horizontal': {
    description: 'The horizontal padding.',
    type: 'string',
    defaultValue: '22px',
  },
  'button.padding.vertical': {
    description: 'The vertical padding.',
    type: 'string',
    defaultValue: '4px',
  },
  'button.extend': {
    description: 'Any additional style for the Button.',
    type: 'string | (props) => {}',
  },
}
