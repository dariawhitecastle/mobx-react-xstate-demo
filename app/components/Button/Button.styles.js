import styled from 'styled-components'
import { Button } from 'grommet'

export const StyledButton = styled(Button)`
  // border-radius: 3px;
`
