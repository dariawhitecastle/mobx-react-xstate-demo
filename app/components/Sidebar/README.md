<!-- ! This is a generated file. To make changes, edit <Component>.doc.js ! -->
## withRouter()
A Sidebar

## Usage

```javascript
import { Sidebar } from 'components/Sidebar';
<Sidebar />
```

## Properties

**routes**

Array of routes to be rendered in the sidebar Defaults to `[
  {
    "label": "Home",
    "exact": true,
    "path": "/"
  }
]`.

```
[{
  label: string,
  exact: boolean,
  path: string,
  main: function
}]
```
  