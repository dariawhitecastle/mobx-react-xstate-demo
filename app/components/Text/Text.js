/* eslint-disable react/prop-types */

import React from 'react'
import { Text } from 'grommet'

import { doc } from './Text.doc'
import helpers from '../../utils/helpers'

/**
 *
 * Text
 *
 */
function createText({
  alignSelf,
  a11yTitle,
  as,
  gridArea,
  margin,
  children,
  color,
  level,
  responsive,
  size,
  textAlign,
  tag,
  truncate,
  weight,
  wordBreak,
  ...rest
}) {
  return (
    <Text
      alignSelf={alignSelf}
      a11yTitle={a11yTitle}
      as={as}
      gridArea={gridArea}
      margin={margin}
      color={color}
      level={level}
      responsive={responsive}
      size={size}
      textAlign={textAlign}
      tag={tag}
      truncate={truncate}
      weight={weight}
      wordBreak={wordBreak}
      {...rest}
    >
      {children}
    </Text>
  )
}

export default helpers.createWithDoc({
  docFunction: doc,
  component: createText,
})
