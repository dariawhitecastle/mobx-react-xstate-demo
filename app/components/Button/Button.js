/* eslint-disable react/prop-types */
/**
 *
 * Button.js
 *
 * A common button
 */

import React, { forwardRef } from 'react'
import { Button } from 'grommet'

import { doc } from './Button.doc'
import helpers from '../../utils/helpers'

const createButton = forwardRef((props, ref) => {
  const {
    a11yTitle,
    active,
    alignSelf,
    as,
    color,
    disabled,
    fill,
    focusIndicator,
    gap,
    gridArea,
    hoverIndicator,
    href,
    icon,
    margin,
    onClick,
    plain,
    primary,
    reverse,
    target,
    title,
    type,
    ...rest
  } = props

  return (
    <Button
      a11yTitle={a11yTitle}
      active={active}
      alignSelf={alignSelf}
      as={as}
      color={color}
      disabled={disabled}
      fill={fill}
      focusIndicator={focusIndicator}
      gap={gap}
      gridArea={gridArea}
      hoverIndicator={hoverIndicator}
      href={href}
      icon={icon}
      label={title}
      margin={margin}
      onClick={onClick}
      plain={plain}
      primary={primary}
      ref={ref}
      reverse={reverse}
      target={target}
      type={type}
      {...rest}
    />
  )
})

export default helpers.createWithDoc({
  docFunction: doc,
  component: createButton,
})
