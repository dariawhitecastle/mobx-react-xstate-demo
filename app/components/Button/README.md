<!-- ! This is a generated file. To make changes, edit <Component>.doc.js ! -->
## createButton
A button.

You can provide a single function child that will be called with
      'hover' and 'focus' keys. This allows you to customize the rendering
      of the Button in those cases.

## Usage

```javascript
import { Button } from 'components/Button';
<Button primary label='Label' />
```

## Properties

**a11yTitle**

Custom title to be used by screen readers.

```
string
```

**alignSelf**

How to align along the cross axis when contained in
      a Box or along the column axis when contained in a Grid.

```
start
center
end
stretch
```

**gridArea**

The name of the area to place
    this inside a parent Grid.

```
string
```

**margin**

The amount of margin around the component. An object can
      be specified to distinguish horizontal margin, vertical margin, and
      margin on a particular side.

```
none
xxsmall
xsmall
small
medium
large
xlarge
{
  bottom: 
    xxsmall
    xsmall
    small
    medium
    large
    xlarge
    string,
  horizontal: 
    xxsmall
    xsmall
    small
    medium
    large
    xlarge
    string,
  left: 
    xxsmall
    xsmall
    small
    medium
    large
    xlarge
    string,
  right: 
    xxsmall
    xsmall
    small
    medium
    large
    xlarge
    string,
  top: 
    xxsmall
    xsmall
    small
    medium
    large
    xlarge
    string,
  vertical: 
    xxsmall
    xsmall
    small
    medium
    large
    xlarge
    string
}
string
```

**active**

Whether the button is active.

```
boolean
```

**color**

Fill color for primary, label color for plain, border color otherwise.

```
string
{
  dark: string,
  light: string
}
```

**disabled**

Whether the button is disabled.

```
boolean
```

**fill**

Whether the button expands to fill all of the available width and/or height.

```
horizontal
vertical
boolean
```

**focusIndicator**

Whether when 'plain' it should receive a focus outline. Defaults to `true`.

```
boolean
```

**hoverIndicator**

The hover indicator to apply when the user is mousing over the
button. An object can be also be specified for color index support:
{background: 'neutral-2'}. This prop is meant to be used only
with plain Buttons.

```
boolean
string
background
{
  background: 
    boolean
    string
}
```

**href**

If specified, the button will behave like an anchor tag.

```
string
```

**target**

Specifies where to display the URL defined in the href property.

```
_self
_blank
_parent
_top
```

**icon**

Icon element to place in the button.

```
element
```

**gap**

The amount of spacing between icon and label in the button. Defaults to `small`.

```
xxsmall
xsmall
small
medium
large
xlarge
string
```

**label**

Label text to place in the button.

```
node
```

**onClick**

Click handler. Not setting this property and not specifying a href
causes the Button to be disabled.

```
function
```

**plain**

Whether this is a plain button with no border or pad.
Non plain button will show both pad and border.
The plain button has no border and unless the icon prop exist it has no pad as well.

```
boolean
```

**primary**

Whether this is a primary button. There should be at most one per page or screen.

```
boolean
```

**reverse**

Whether an icon and label should be reversed so that the icon is at the
end of the anchor.

```
boolean
```

**type**

The type of button. Set the type to submit for the default button on forms. Defaults to `button`.

```
button
reset
submit
```

**as**

The DOM tag or react component to use for the element.

```
string
function
```
  
## Intrinsic element

```
button
```
## Theme
  
**global.hover.color**

The background color when hovering. Expects `string`.

Defaults to

```
{ dark: 'white', light: 'black' }
```

**global.edgeSize.small**

The padding around an icon-only button. Expects `string`.

Defaults to

```
12px
```

**global.colors.control**

The color of the border. Expects `string | { dark: string, light: string }`.

Defaults to

```
{ dark: 'accent-1', light: 'brand', }
```

**global.colors.brand**

The light version of the border. Expects `string`.

Defaults to

```
#7D4CDB
```

**global.colors.text**

The color of the text label. Expects `string | { dark: string, light: string }`.

Defaults to

```
{ dark: '#f8f8f8', light: '#444444' }
```

**text.medium.size**

The font size of the text label. Expects `string`.

Defaults to

```
18px
```

**text.medium.height**

The line height of the text label. Expects `string`.

Defaults to

```
24px
```

**button.border.color**

The color of the border. Expects `string | { dark: string, light: string }`.

Defaults to

```
undefined
```

**button.border.radius**

The corner radius. Expects `string`.

Defaults to

```
18px
```

**button.border.width**

The border width. Expects `string`.

Defaults to

```
2px
```

**button.color**

The color of the text label. Expects `string | { dark: string, light: string }`.

Defaults to

```
undefined
```

**button.primary.color**

The color of the background for primary buttons. Expects `string | { dark: string, light: string }`.

Defaults to

```
undefined
```

**button.disabled.opacity**

The opacity when the button is disabled. Expects `number`.

Defaults to

```
0.3
```

**button.padding.horizontal**

The horizontal padding. Expects `string`.

Defaults to

```
22px
```

**button.padding.vertical**

The vertical padding. Expects `string`.

Defaults to

```
4px
```

**button.extend**

Any additional style for the Button. Expects `string | (props) => {}`.

Defaults to

```
undefined
```
