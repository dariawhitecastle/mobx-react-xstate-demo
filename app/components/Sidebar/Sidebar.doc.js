import React from 'react'
import { describe, PropTypes } from 'react-desc'
import { Button } from '../Button'

export const doc = Sidebar => {
  const DocumentedSidebar = describe(Sidebar)
    .description('A Sidebar')
    .usage(
      `import { Sidebar } from 'components/Sidebar';
<Sidebar />`,
    )

  DocumentedSidebar.propTypes = {
    routes: PropTypes.arrayOf(
      PropTypes.shape({
        label: PropTypes.string.isRequired,
        exact: PropTypes.bool,
        path: PropTypes.string.isRequired,
        main: PropTypes.func.isRequired,
      }),
    )
      .description('Array of routes to be rendered in the sidebar')
      .defaultValue([
        {
          label: 'Home',
          exact: true,
          path: '/',
          main: () => <Button label="Click me" />,
        },
      ]),
  }

  return DocumentedSidebar
}
