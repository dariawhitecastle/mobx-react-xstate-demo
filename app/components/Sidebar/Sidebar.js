import React, { useState } from 'react'
import PropTypes from 'prop-types'
import { Link, withRouter } from 'react-router-dom'

import { Box } from '../Box'
import { Button } from '../Button'
import { Text } from '../Text'

import colors from '../../utils/colors'
import { fontSizes } from '../../utils/sizes'
import helpers from '../../utils/helpers'

import { doc } from './Sidebar.doc'

export const SidebarButton = ({ title, path, ...rest }) => (
  <Button plain {...rest}>
    {({ hover }) => (
      <Link style={{ textDecoration: 'none' }} to={path}>
        <Box
          align="start"
          background={hover || rest.active ? colors.primaryBlue50 : undefined}
          pad={{ horizontal: '15px', vertical: '10px' }}
          border={{ color: colors.primaryBlue50, size: 'xsmall', side: 'bottom' }}
        >
          <Text
            weight={rest.active ? 'bold' : 'normal'}
            size={fontSizes.small}
            color={colors.primaryText}
          >
            {title}
          </Text>
        </Box>
      </Link>
    )}
  </Button>
)

const Sidebar = withRouter(({ routes, location }) => {
  const currentActive = routes.find(route => route.path === location.pathname)
  const [active, setActive] = useState(currentActive.label)
  return (
    <Box background={colors.primaryBlue10} fill>
      {routes.map(({ label, path }) => (
        <SidebarButton
          key={label}
          title={label}
          path={path}
          active={label === active}
          onClick={() => setActive(label)}
        />
      ))}
    </Box>
  )
})

export default helpers.createWithDoc({
  docFunction: doc,
  component: Sidebar,
})

SidebarButton.propTypes = {
  title: PropTypes.string.isRequired,
  path: PropTypes.string.isRequired,
}

Sidebar.propTypes = {
  routes: PropTypes.arrayOf(
    PropTypes.shape({
      label: PropTypes.string.isRequired,
      exact: PropTypes.bool,
      path: PropTypes.string.isRequired,
      main: PropTypes.func.isRequired,
    }),
  ),
}
