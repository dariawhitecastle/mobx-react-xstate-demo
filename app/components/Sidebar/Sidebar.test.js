/**
 *
 * Tests for Sidebar
 *
 * @see https://github.com/react-boilerplate/react-boilerplate/tree/master/docs/testing
 *
 */

import React from 'react'
import { render } from '@testing-library/react'
import { MemoryRouter } from 'react-router-dom'
// import 'jest-dom/extend-expect'; // add some helpful assertions

import { Button } from '../Button'
import { Text } from '../Text'

import Sidebar from './Sidebar'

const routes = [
  {
    label: 'Hello',
    exact: true,
    path: '/hello',
    main: () => <Text alignSelf="center">Hello!</Text>,
  },
  {
    label: 'Bananas',
    path: '/bananas',
    main: () => <Button label="Click me!" onClick={() => console.info('Bananas!')} />,
  },
]

const renderComponent = (props = {}) =>
  render(
    <MemoryRouter initialEntries={['/bananas']}>
      <Sidebar {...props} routes={routes} />
    </MemoryRouter>,
  )

test('it renders and matches snapshot', () => {
  const { container } = renderComponent()
  expect(container).toMatchSnapshot()
})

test('It does not log to the console', () => {
  const spy = jest.spyOn(global.console, 'log')
  renderComponent()
  expect(spy).not.toHaveBeenCalled()
})
