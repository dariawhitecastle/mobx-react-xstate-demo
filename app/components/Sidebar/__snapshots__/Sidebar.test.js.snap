// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it renders and matches snapshot 1`] = `
<div>
  <div
    class="StyledBox-sc-13pk1d4-0 eTzjKl"
  >
    <button
      class="StyledButton-sc-323bzc-0 jfbKsv"
      type="button"
    >
      <a
        href="/hello"
        style="text-decoration: none;"
      >
        <div
          class="StyledBox-sc-13pk1d4-0 iURajA"
        >
          <span
            class="StyledText-sc-1sadyjn-0 cfieNy"
          >
            Hello
          </span>
        </div>
      </a>
    </button>
    <button
      class="StyledButton-sc-323bzc-0 bAOLDb"
      type="button"
    >
      <a
        href="/bananas"
        style="text-decoration: none;"
      >
        <div
          class="StyledBox-sc-13pk1d4-0 eedNja"
        >
          <span
            class="StyledText-sc-1sadyjn-0 kVyNC"
          >
            Bananas
          </span>
        </div>
      </a>
    </button>
  </div>
</div>
`;
